Add API for working with rules

psybers opened the pull request:
This adds the following API endpoints for working with payee rules:

- getPayeeRules()
- createPayeeRule(rule)
- updatePayeeRule(id, rule)
- deletePayeeRule(id)

Note that `deletePayeeRule` accepts either the rule ID *or* a rule object (that has an `id` property).  So you can pass in things returned from `getPayeeRules()` there.

`updatePayeeRule` will ignore the `id` property on the rule, favoring the one passed in as first argument.

psybers commented:
Note that once this is approved, I will go and revert+revise this docs commit: https://github.com/actualbudget/docs/commit/90c454450f699f8060edf2f888077c22a6c4a621

joel-jeremy commented on packages/loot-core/src/types/server-handlers.d.ts:
```diff
@@ -118,7 +118,13 @@ export interface ServerHandlers {
   'payees-check-orphaned': (arg: { ids }) => Promise<unknown>;
 
-  'payees-get-rules': (arg: { id }) => Promise<unknown>;
+  'payees-get-rules': (arg: { id }) => Promise<unknown[]>;
```
Can we use RuleEntity here?

psybers replied:
Maybe? I just fixed the arity issue. ;)

joel-jeremy commented on packages/loot-core/src/server/api.ts:
```diff
@@ -600,6 +600,26 @@ handlers['api/payee-delete'] = withMutation(async function ({ id }) {
+  return handlers['payee-rule-update']({ id, rule });
+});
+
+handlers['api/payee-rule-delete'] = withMutation(async function (id) {
```
I'm leaning towards making this just accept an object with an id for consistency with the other delete methods we have.

joel-jeremy commented:
Does these methods only support rules with payee as condition?

psybers commented:
> Does these methods only support rules with payee as condition?

I *think* they support any rule. But the getter might be payee-only rules? I just reused the existing getter.

I wasn't entirely sure what we wanted here.  The create/update/delete methods could be renamed to remove 'payee' from their names, because those definitely work on any rule.

Maybe we do that and also add two getters: one for rules for a given payee, and one for just all rules?

joel-jeremy commented:
> Maybe we do that and also add two getters: one for rules for a given payee, and one for just all rules?

Agreed :) We should just retain the existing getter for payee and add a generic rules getter.

psybers commented on packages/loot-core/src/server/api.ts:
```diff
@@ -600,6 +600,32 @@ handlers['api/payee-delete'] = withMutation(async function ({ id }) {
+handlers['api/rule-create'] = withMutation(async function ({ rule }) {
+  checkFileOpen();
+  const addedRule = await handlers['rule-add'](rule);
+  return (addedRule as { id }).id;
```
Since the other create functions seem to just return the string ID, I tried to mimic that here.  But that means if there was an `error` set, not only is that error message not propagated to the client but now this type cast should throw a different error.  This probably is not what we want, but I was not sure what to do here.

MatissJanis replied:
Anything we could do here to remove the manual type coercion?

psybers commented on packages/api/methods.test.ts:
```diff
@@ -346,6 +346,166 @@ describe('API CRUD operations', () => {
+        },
+      ],
+    };
+    const ruleId = await api.createRule(rule);
+    rule.id = ruleId;
```
This feels super weird.  I had to add an `id: null` into the object when creating it and then patch the ID back in.  But as far as I can tell, the other creates just return a string ID so this is mimicking that behavior.

psybers commented:
@joel-jeremy I updated quite a bit. Discovered there were already handlers for create/update/delete, so removed mine and re-used those.  Renamed to remove `payee` from those names.  Made two getters: one for all rules, one for payee rules.

psybers commented on packages/api/methods.test.ts:
```diff
@@ -346,6 +346,221 @@ describe('API CRUD operations', () => {
+    );
+
+    // get by payee
+    rules = await api.getPayeeRules(payeeId1);
```
Currently, these calls return an empty list, instead of the matching rule(s).

MatissJanis commented on packages/loot-core/src/types/api-handlers.d.ts:
```diff
@@ -143,4 +143,14 @@ export interface ApiHandlers {
+
+  'api/rules-get': () => Promise<RuleEntity[]>;
+
+  'api/payee-rules-get': (arg: { id }) => Promise<RuleEntity[]>;
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏missing argument type (here and also the other API methods below)
```suggestion
  'api/payee-rules-get': (arg: { id: string }) => Promise<RuleEntity[]>;
```

MatissJanis commented on packages/loot-core/src/types/server-handlers.d.ts:
```diff
@@ -118,7 +119,7 @@ export interface ServerHandlers {
   'payees-check-orphaned': (arg: { ids }) => Promise<unknown>;
 
-  'payees-get-rules': (arg: { id }) => Promise<unknown>;
+  'payees-get-rules': (arg: { id }) => Promise<RuleEntity[]>;
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏I know this is not new code, but if we're touching it.. might as well improve it even further :)

```suggestion
  'payees-get-rules': (arg: { id: string }) => Promise<RuleEntity[]>;
```

MatissJanis reviewed (commented):
Just a few small nitpicks. Overall I think this makes sense. Thanks for working on adding these APIs!

MatissJanis commented:
LGTM from my end by the way. But I don't want to approve & merge myself as I contributed here too. So ideally someone neutral can review :)

youngcw reviewed (dismissed):
That all seems fine to me

youngcw commented:
@psybers Would you be willing to add these functions to the api docs page?

psybers commented:
> @psybers Would you be willing to add these functions to the api docs page?

Yes. My intent was to revert a prior commit that removed those docs and then update them to match what we have here.

youngcw commented:
@psybers could you fix these merge conflicts so we can merge this in?

psybers commented:
@youngcw said:
> @psybers could you fix these merge conflicts so we can merge this in?

Fixed.

psybers commented:
Documentation work will be here: https://github.com/actualbudget/docs/pull/356
